src/jit/morph.cpp

```diff
@@ -9315,232 +9314,260 @@ GenTree* Compiler::fgMorphInitBlock(GenTree* tree)
+            if (!dest->OperIsBlk())
+            {
+                noway_assert(blockSize != 0);
+                tree->gtOp.gtOp1 = origDest;
+                tree->gtType     = origDest->gtType;
+            }
+
+            dest             = fgMorphBlockOperand(dest, dest->TypeGet(), blockSize, true);
             tree->gtOp.gtOp1 = dest;
             tree->gtFlags |= (dest->gtFlags & GTF_ALL_EFFECT);
         }
-        else
-        {
-            // The initVal must be a constant of TYP_INT
-            noway_assert(initVal->OperGet() == GT_CNS_INT);
-            noway_assert(genActualType(initVal->gtType) == TYP_INT);
+    }
 
-            // The dest must be of a struct type.
-            noway_assert(varTypeIsStruct(destLclVar));
+#ifdef DEBUG
+    if (morphed)
+    {
+        tree->gtDebugFlags |= GTF_DEBUG_NODE_MORPHED;
 
-            //
-            // Now, convert InitBlock to individual assignments
-            //
+        if (verbose)
+        {
+            printf("fgMorphInitBlock (after):\n");
+            gtDispTree(tree);
+        }
+    }
+#endif
 
-            tree = nullptr;
-            INDEBUG(morphed = true);
+    return tree;
+}
 
-            GenTree* dest;
-            GenTree* srcCopy;
-            unsigned fieldLclNum;
-            unsigned fieldCnt = destLclVar->lvFieldCnt;
+//------------------------------------------------------------------------
+// fgMorphPromoteLocalInitBlock: Attempts to promote a local block init tree
+// to a tree of promoted field initialization assignments.
+//
+// Arguments:
+//    destLclNode - The destination LclVar node
+//    initVal - The initialization value
+//    blockSize - The amount of bytes to initialize
+//
+// Return Value:
+//    A tree that performs field by field initialization of the destination
+//    struct variable if various conditions are met, nullptr otherwise.
+//
+GenTree* Compiler::fgMorphPromoteLocalInitBlock(GenTreeLclVar* destLclNode, GenTree* initVal, unsigned blockSize)
+{
+    assert(destLclNode->OperIs(GT_LCL_VAR));
 
-            for (unsigned i = 0; i < fieldCnt; ++i)
-            {
-                fieldLclNum = destLclVar->lvFieldLclStart + i;
-                dest        = gtNewLclvNode(fieldLclNum, lvaTable[fieldLclNum].TypeGet());
+    LclVarDsc* destLclVar = lvaGetDesc(destLclNode);
+    assert(varTypeIsStruct(destLclVar->TypeGet()));
+    assert(destLclVar->lvPromoted);
 
-                noway_assert(lclVarTree->gtOper == GT_LCL_VAR);
-                // If it had been labeled a "USEASG", assignments to the the individual promoted fields are not.
-                dest->gtFlags |= (lclVarTree->gtFlags & ~(GTF_NODE_MASK | GTF_VAR_USEASG));
+    if (blockSize == 0)
+    {
+        JITDUMP(" size is not known.\n");
+        return nullptr;
+    }
 
-                srcCopy = gtCloneExpr(initVal);
-                noway_assert(srcCopy != nullptr);
+    if (destLclVar->lvAddrExposed && destLclVar->lvContainsHoles)
```

Comment:
> Could you please give an example of initializing struct with holes overlapping fields?

Such a struct can't ever be promoted, you'd loose the interference between the overlapped fields (unless you modify the IR in such a way to preserve the interference, that would probably be pretty crazy). `CanPromoteStructType` checks this.

So you get normal block initialization:
```C#
Unsafe.InitBlock(&ms, 42, 8);
```
generates
```asm
48B92A2A2A2A2A2A2A2A mov      rcx, 0x2A2A2A2A2A2A2A2A
48894C2428           mov      qword ptr [rsp+28H], rcx
```

> Is not destLclVar->lvContainsHoles enough to reject this optimization? As we saw from #18988 we can access holes even if struct was not defined as destLclVar->lvAddrExposed (via Unsafe.As<>).

Well, that's just how the existing code works, I just moved it around because it was terribly confusing, with similar conditions being tested in different places for no good reason. 

I'd say that it's the right thing to do:
* Rejecting field by field initialization just because the struct has holes is overly conservative - various mixes of integer field types will naturally cause holes due to alignment requirements.
* Normally nobody cares about what values holes contain. They might care when holes are created on purpose (e.g. with explicit layout) but that case is rejected below.
* It may be considered a security issue if it allows information to leak through holes but with CAS being practically dead that's unlikely.